### ISSUE TITLE
[Bug] MaskedBehavior Cursor jumps to start on unmasked Characters

### ISSUE THREAD

**Issue by sh-kd** [CommunityToolkit/Maui#460]
### Description

The cursor jumps to the start when an maskedcharacter is reached.

### Link to Reproduction Sample

**https://github.com/sh-kd/MaskedBehaviorBug**

### Steps to Reproduce

1.  Make an entry with MaskedBehavior that has an maskedcharacter
2.  Type in the entry till you reach the maskedcharacter

### Expected Behavior

That the Cursor goes trough the maskedcharacters

### Actual Behavior

The Cursor jumps to the start of the entry. So you have to move the cursor to the end every time an masked character is reached.

### Basic Information

- Version with issue: 1
- Last known good version:
- IDE: VS2022
- Platform Target Frameworks: Android

- Android Support Library Version: 
- Nuget Packages: MAUI Toolkit
- Affected Devices: Honeywell RT10A      Honeywell TA60A     Android 10 API 29. Dont know if it affects other devices

### Workaround

no

### Reproduction imagery

I cant

**Comment by VictorHFerreira** [comment]
i have the same issue here and i dont think it's from maui. My entry was working fine, until i put masked behavior on it. On windows it works fine, but in android, tha issue happens

**Comment by EthanHipps** [comment]
I'm also experiencing this issue in my project (maui v7.0.49 and toolkit v8.0.0). Is there any update on what's causing this since the linked maui issue, [#6916](https://github.com/dotnet/maui/issues/6916), is now closed?

**Comment by EPS-Lac** [comment]
I've got the same issue, also with maui v7.0.49 and toolkit.maui v3.0.0.
I'm using a Samsung Galaxy A53, Samsung Keyboard Settings (English US), in case this might be a device related problem.

**Comment by VladislavAntonyuk** [comment]
Some code changes ITextInput CursorPosition value right after adding a second character after the mask (e.g. MASK 9999-9999-9999-9999 Expected Input 1234-5678-1234-5678 Actual Input 1234-5 (When I click 6 I get result 1623-45)

**Comment by KieranMaclagan** [comment]
Seems to be Android only. Works as expected on iOS (both simulator and physical), but getting the issue on a physical Pixel 6.

**Comment by FM1973** [comment]
I got the same problem on Android.

**Comment by Asfiroth** [comment]
Any workaround for this issue?, I'm having the same problem on Android

**Comment by anotherlab** [comment]
I'm seeing the same problem on Android, specifically with the emulator and WSA. Appear to work fine with iOS.

**Comment by Asfiroth** [comment]
Yeah, I opened an [issue](https://github.com/dotnet/maui/issues/13571) on MAUI repo for this, and apparently is fixed on .net-8, but it might not get back-ported to .net-7 😞

**Comment by insolito86** [comment]
Anyone got it to work on .net 7?

**Comment by Asfiroth** [comment]
Hi @insolito86, on the issue I put a workaround for this case, as I say there is not perfect but works.

**Comment by candidodmv** [comment]
Hello @Asfiroth.
Wich workaround have you used? could you share?

Thank you.

**Comment by Asfiroth** [comment]
Hi @candidodmv sure:
modifying the Entry handler for android adding this:

On ConnectHandler:

`PlatformView.TextChanged += PlatformViewOnTextChanged;`

On DisconnectHandler:

`PlatformView.TextChanged -= PlatformViewOnTextChanged;`

This is the event handler for text changed event
```

private void PlatformViewOnTextChanged(object sender, TextChangedEventArgs e)
    {
        if (sender is not AppCompatEditText editText) return;
        
        if (string.IsNullOrWhiteSpace(editText.Text)) return;
        
        var looper = Looper.MyLooper();
        if (looper != null)
        {
            var handler = new Handler(looper);
            handler.Post(() =>
            {
                editText.SetSelection(editText.Text.Length);
            });
        }else
        {
            editText.SetSelection(editText.Text.Length);
        }
    }
```

As I said, It's not perfect but does the job, any help to make this better is appreciated.

**Comment by VictorHFerreira** [comment]
what worked for me was in text changed for the entry, add the following code

(sender as Entry).CursorPosition = (sender as Entry).Text.Length;

**Comment by candidodmv** [comment]
Hello @Asfiroth, Thank you for your effort in helping!

But I did preferer to return to v6.0 version v7.0 is pretty unstable yet. At version 6.0 things work as expected.

**Comment by NunoBem** [comment]
I still have this issue in the latest .NET 7, is it really fixed in .NET 8 (can't install preview on work machine)?

**Comment by LarryOlsen** [comment]
I am seeing this bug on .NET 8 RC2

**Comment by aelmardi** [comment]
I confirm that it does not work on the GA release of .NET 8.0.3 either.

**Comment by DaveEvans1968** [comment]
Either remove the function from the community toolkit or fix it.

**Comment by brminnick** [comment]
@DaveEvans1968 If you have a solution, please submit it via a Pull Request!

**Comment by vikher** [comment]
I'm seeing this bug in the .NET 8.0.3 .
This is a workaround for a problem where the cursor jumps to the start when using certain behavior in the code.
To fix it, in the file called Page.xaml.cs, include the following code:

```
void OnEntryHandlerChanged(object sender, EventArgs e)
{
    // Only for Android
    if (sender is CnkEntryView entryView)
    {
        Entry entry = entryView.Entry;
        (entry.Handler.PlatformView as AppCompatEditText).TextChanged += textChangedEventHandler;
    }
}

void OnEntryHandlerChanging(object sender, HandlerChangingEventArgs e)
{
    // Only for Android
    if (e.OldHandler != null)
    {
        (e.OldHandler.PlatformView as AppCompatEditText).TextChanged -= textChangedEventHandler;
    }
}

#if ANDROID
private void textChangedEventHandler(object sender, Android.Text.TextChangedEventArgs e)
{
    if (!(sender is AppCompatEditText editText) || string.IsNullOrWhiteSpace(editText.Text))
        return;

    try
    {
        var looper = Looper.MyLooper();
        var handler = new Handler(looper ?? Looper.MainLooper);
        handler.Post(() => editText.SetSelection(editText.Text.Length));
    }
    catch (Exception ex)
    {
        MobileApp.Instance.LogException(ex);
    }
}
#endif

```

Additionally, in the Page.xaml file, use the following code:

```
<customElements:EntryView x:Name="phoneEntry"
                 HandlerChanged="OnEntryHandlerChanged"
                 HandlerChanging="OnEntryHandlerChanging"
                 LabelText="Phone"
                 Keyboard="Numeric"                
                 BindingContext="{Binding PhoneNumber}"
                 Converter="{StaticResource PhoneNumberConverter}"
                 Margin="0,20,0,0"/>
```

**Comment by kimhongka** [comment]
I am on .NET 8.0.14 and still having this issue. Please fix it

**Comment by brminnick** [comment]
@kimhongka Would you like to submit a Pull Request?

**Comment by kimhongka** [comment]
The only way that is working for me is

<img width="655" alt="Screenshot 2024-04-04 at 11 34 32 AM" src="https://github.com/CommunityToolkit/Maui/assets/2062330/07a4e3ca-e33f-4492-826f-505183dc7d41">


I've tried to use Entry Handler way and also adding 

    protected override void OnTextChanged(string oldValue, string newValue)
        {
            CursorPosition = Text.Length;
        }
        with extended entry, but it does not even trigger and both handler and the override method blocks the MaskedBehavior. 
        
        I will go with the workaround solution for now.

**Comment by Hackmodford** [comment]
I am still experiencing this in .NET 8

**Comment by NunoBem** [comment]
> I am still experiencing this in .NET 8

Me too. Who is experiencing this should put an emoji on the issue, so it gets noticed.

**Comment by bijington** [comment]
> > I am still experiencing this in .NET 8
> 
> 
> 
> Me too. Who is experiencing this should put an emoji on the issue, so it gets noticed.

Or feel free to submit a PR to fix the issue

**Comment by NunoBem** [comment]
Hi Shaun,
Wish I had the knowledge for this. Do we know is an issue with the toolkit and not maui itself?

**Comment by bijington** [comment]
I don't know. The history of this or similar issues implies the issue is in MAUI but I couldn't say for sure

**Comment by b12kab** [comment]
I think they are different, as I can work around the cursor problem here (toolkit), but cannot in Maui. For Maui, I had a workaround which moved the cursor to the correct position, but it immediately moves it back to the incorrect position.

My workaround:  https://b12kab@bitbucket.org/b12kab/communitytoolkitmaskedbehaviorworkaround.git

**Comment by bijington** [comment]
I read that as the workaround doesn't solve this issue, is that correct? If I am wrong and it does solve the issue is there something that could be submitted to the toolkit to fix this issue?

**Comment by b12kab** [comment]
Correct, it just works around the exiting issue; it does not solve it.

**Comment by Hackmodford** [comment]
> Yeah, I opened an [issue](https://github.com/dotnet/maui/issues/13571) on MAUI repo for this, and apparently is fixed on .net-8, but it might not get back-ported to .net-7 😞

It is not fixed in .NET 8

**Comment by Asfiroth** [comment]
Hi @Hackmodford , it appears so, I encourage you to open a new issue on MAUI repo since the one I opened is now closed by the team.

**Comment by Hackmodford** [comment]
@Asfiroth Done
https://github.com/dotnet/maui/issues/22497

**Comment by LeoJHarris** [comment]
> what worked for me was in text changed for the entry, add the following code
> 
> (sender as Entry).CursorPosition = (sender as Entry).Text.Length;

Work around of sorts but cant update text in the middle of the string without the cursor jumping to the end all the time, seems to be similar issue plaguing the mask behavior from Xamarin days.....

**Comment by SamDean** [comment]
FWIW, I just updated from Maui 8.0.40 => 8.0.70 and CommunityToolkit 9.0.0 => 9.0.2 
and I no longer have the issue! Happy days.

**Comment by marcelozanetti** [comment]
> FWIW, I just updated from Maui 8.0.40 => 8.0.70 and CommunityToolkit 9.0.0 => 9.0.2 and I no longer have the issue! Happy days.

337 / 5.000
I downgraded to the version you suggested, it worked partially. When I try to change any digit in the middle of the numbering, before any mask character, the cursor is thrown to the last position of the element. If I update MAUI to the latest version (8.0.92) and the latest Maui ToolKit (9.1.0), it gets even worse.

**Comment by LeoJHarris** [comment]
This definitely not fixed. I have the same issues as @marcelozanetti that it still jumps to the end depending on where you are in the string editing text 

Consider the following mask, if you edit within the last 3 character `XXX` then it works fine but editing text before the area in bold XXXX XXX **XXX** results in the jump.

`XXXX XXX XXX`

This issue should be reopened. Did anyway even confirm the issue is resolved with thoroughly testing different scenarios?

**Comment by bijington** [comment]
@LeoJHarris if you look at the PR at was aimed at fixing this issue I wrote that I believed my change was more a workaround than a proper fix. That said I reproduced and then verified the change solved the issue based on the limited scenario that was presented. 

If you have other scenarios where you are seeing this issue or perhaps there is a regression due to underlying changes in MAUI then I will kindly ask you to open a new issue with a sample reproduction of your scenarios. Feel free to refer back to this issue
